src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+        private Task assignerLoop;
+
+        /// <summary>Loop that checks if peers failed to deliver important blocks in given time and penalizes them if they did.</summary>
+        private Task stallingLoop;
+
+        public BlockPuller(OnBlockDownloadedCallback callback, IChainState chainState, ProtocolVersion protocolVersion, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        {
+            this.reassignedJobsQueue = new Queue<DownloadJob>();
+            this.downloadJobsQueue = new Queue<DownloadJob>();
+
+            this.assignedDownloadsByHash = new Dictionary<uint256, AssignedDownload>();
+            this.assignedDownloadsSorted = new LinkedList<AssignedDownload>();
+            this.assignedHeadersByPeerId = new Dictionary<int, List<ChainedHeader>>();
+
+            this.averageBlockSizeBytes = new AverageCalculator(AverageBlockSizeSamplesCount);
+            
+            this.pullerBehaviorsByPeerId = new Dictionary<int, IBlockPullerBehavior>();
+
+            this.processQueuesSignal = new AsyncManualResetEvent(false);
+            this.queueLock = new object();
+            this.peerLock = new object();
+            this.assignedLock = new object();
+            this.nextJobId = 0;
+            
+            this.networkPeerRequirement = new NetworkPeerRequirement
+            {
+                MinVersion = protocolVersion,
+                RequiredServices = NetworkPeerServices.Network
+            };
+
+            this.cancellationSource = new CancellationTokenSource();
+            this.random = new Random();
+
+            this.maxBlocksBeingDownloaded = MinimalCountOfBlocksBeingDownloaded;
+
+            this.OnDownloadedCallback = callback;
+            this.chainState = chainState;
+            this.dateTimeProvider = dateTimeProvider;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc/>
+        public void Initialize()
+        {
+            this.logger.LogTrace("()");
+
+            this.assignerLoop = this.AssignerLoopAsync();
+            this.stallingLoop = this.StallingLoopAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public double GetAverageBlockSizeBytes()
+        {
+            lock (this.queueLock)
+            {
+                return this.averageBlockSizeBytes.Average;
+            }
+        }
+
+        private int GetTotalSpeedOfAllPeersBytesPerSec()
+        {
+            lock (this.peerLock)
+            {
+                return this.pullerBehaviorsByPeerId.Sum(x => x.Value.SpeedBytesPerSecond);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void OnIbdStateChanged(bool isIbd)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(isIbd), isIbd);
+
+            lock (this.peerLock)
+            {
+                foreach (IBlockPullerBehavior blockPullerBehavior in this.pullerBehaviorsByPeerId.Values)
```

Comment:
Arent you suppose to also change `averageBlockSizeBytes`?

Reply:
No. IBD state doesn't affect the max amount of samples for averageBlockSizeBytes AverageCalculator